packages/mds-event-processor/src/proc-event.ts

```diff
@@ -0,0 +1,269 @@
+  //TODO reduce logic
+  if (trip_id) {
+    if (!trips[trip_id]) {
+      trips[trip_id] = []
+    }
+    trips[trip_id].push(tripEvent)
+  }
+
+  // Update trip event cache and stream
+  await cache.hset('trips:events', provider_id + ':' + device_id, JSON.stringify(trips))
+  //await stream.writeCloudEvent('mds.trip.event', JSON.stringify(tripEvent))
+  await processTripTelemetry(deviceState)
+}
+
+async function getTripId(deviceState: StateEntry) {
+  /*
+  Return trip_id for telemetery entry by associating timestamps
+  */
+  const { provider_id, device_id, timestamp } = deviceState
+  const tripsEvents: { [trip_id: string]: TripEvent[] } = JSON.parse(
+    await cache.hget('trips:events', provider_id + ':' + device_id)
+  )
+  if (!tripsEvents) {
+    log.warn('NO TRIP DATA FOUND')
+    return undefined
+  } else {
+    let latestStartTime
+    let matchedID
+    // Find latest trip whose start time is before current timestamp
+    // TODO reduce logic with better filtering/mapping ({trip_id, start_time}[])
+    for (let trip_id in tripsEvents) {
+      const tripEvents: TripEvent[] = tripsEvents[trip_id]
+      const startEvents: TripEvent[] = tripEvents.filter(function(tripEvents) {
+        return tripEvents.event_type === 'trip_start' || tripEvents.event_type === 'trip_enter'
+      })
+      startEvents.sort((a, b) => (b.timestamp > a.timestamp ? 1 : -1))
+      const tripStartTime = startEvents[0].timestamp
+      if (tripStartTime <= timestamp && (!latestStartTime || latestStartTime <= tripStartTime)) {
+        matchedID = trip_id
+        latestStartTime = tripStartTime
+      }
+    }
+    if (!matchedID) {
+      log.warn('NO TRIPS MATCHED')
+      return undefined
+    }
+    return matchedID
+  }
+}
+
+async function processTripTelemetry(deviceState: StateEntry) {
+  /*
+    Add trip related telemetry to cache (trips:telemetry):
+
+      Key: 'provider_id:device_id'
+      Value: hash map of tripTelemetry keyed by trip_id
+
+  */
+  const {
+    type,
+    timestamp,
+    annotation_version,
+    annotation,
+    gps,
+    service_area_id,
+    provider_id,
+    device_id,
+    trip_id
+  } = deviceState
+
+  const lat = gps ? gps.lat : null
+  const lng = gps ? gps.lng : null
+  const tripTelemetry = {
+    timestamp: timestamp,
+    latitude: lat,
+    longitude: lng,
+    annotation_version: annotation_version,
+    annotation: annotation,
+    service_area_id: service_area_id
+  } as TripTelemetry
+
+  // Check if associated to an event or telemetry post
+  const tripId = type === 'telemetry' ? getTripId(deviceState) : trip_id
+  if (typeof tripId === 'undefined') {
+    return false
+  }
+  const cacheEntry: string | null = await cache.hget('trips:telemetry', provider_id + ':' + device_id)
+  let trips: { [trip_id: string]: TripTelemetry[] } = cacheEntry ? JSON.parse(cacheEntry) : {}
+
+  //TODO reduce logic
+  if (typeof tripId === 'string') {
```

Comment:
This looks almost identical to a bit of code above, we can probably consolidate into a helper method.